anta/reporter/md_reporter.py

```diff
@@ -0,0 +1,332 @@
+        """Generate and write the various sections of the markdown report.
+
+        Parameters
+        ----------
+            results: The ResultsManager instance containing all test results.
+            md_filename: The path to the markdown file to write the report into.
+            only_failed_tests: Flag to generate a report with only failed tests. Defaults to False.
+        """
+        MDReportBase.ONLY_FAILED_TESTS = only_failed_tests
+
+        try:
+            with md_filename.open("w", encoding="utf-8") as mdfile:
+                sections: list[MDReportBase] = [
+                    ANTAReport(mdfile, results),
+                    TestResultsSummary(mdfile, results),
+                    SummaryTotals(mdfile, results),
+                    SummaryTotalsDeviceUnderTest(mdfile, results),
+                    SummaryTotalsPerCategory(mdfile, results),
+                    FailedTestResultsSummary(mdfile, results),
+                    AllTestResults(mdfile, results),
+                ]
+                for section in sections:
+                    section.generate_section()
+        except OSError as exc:
+            message = f"OSError caught while writing the Markdown file '{md_filename.resolve()}'."
+            anta_log_exception(exc, message, logger)
+            raise
+
+
+class MDReportBase(ABC):
+    """Base class for all sections subclasses.
+
+    Every subclasses must implement the `generate_section` method that uses the `ResultManager` object
+    to generate and write content to the provided markdown file.
+    """
+
+    ONLY_FAILED_TESTS: ClassVar[bool] = False
+
+    def __init__(self, mdfile: TextIOWrapper, results: ResultManager) -> None:
+        """Initialize the MDReportBase with an open markdown file object to write to and a ResultManager instance.
+
+        Parameters
+        ----------
+            mdfile: An open file object to write the markdown data into.
+            results: The ResultsManager instance containing all test results.
+        """
+        self.mdfile = mdfile
+        self.results = results
+
+    @abstractmethod
+    def generate_section(self) -> None:
+        """Abstract method to generate a specific section of the markdown report.
+
+        Must be implemented by subclasses.
+        """
+        msg = "Must be implemented by subclasses"
+        raise NotImplementedError(msg)
+
+    def generate_rows(self) -> Generator[str, None, None]:
+        """Generate the rows of a markdown table for a specific report section.
+
+        Subclasses can implement this method to generate the content of the table rows.
+        """
+        msg = "Subclasses should implement this method"
+        raise NotImplementedError(msg)
+
+    def generate_heading_name(self) -> str:
+        """Generate a formatted heading name based on the class name.
+
+        Returns
+        -------
+            str: Formatted header name.
+
+        Example:
```

Comment:
```suggestion
        Example
```

Reply:
Fixed in 8918cd0822dee86db2c1574d30732c0418bfefff